plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -388,56 +525,245 @@ fu_rts54hub_rtd21xx_device_write_firmware (FuDevice *device,
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error)) {
+			fu_rts54hub_rtd21xx_device_exit_background (self, error);
+			return FALSE;
+		}
+		if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+							   UC_ISP_SLAVE_ADDR,
+							   UC_BACKGROUND_ISP_DATA_OPCODE,
+							   fu_chunk_get_data (chk),
+							   fu_chunk_get_data_sz (chk),
+							   error)) {
+			g_prefix_error (error,
+					"failed to write @0x%04x: ",
+					fu_chunk_get_address (chk));
+			fu_rts54hub_rtd21xx_device_exit_background (self, error);
+			return FALSE;
+		}
+
+		/* update progress */
+		fu_device_set_progress_full (device, (gsize) i, (gsize) chunks->len);
+	}
+
+	/* update finish command */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error)) {
+		fu_rts54hub_rtd21xx_device_exit_background (self, error);
+		return FALSE;
+	}
+	write_buf[0] = ISP_BG_CMD_FW_UPDATE_ISP_DONE;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_ISP_SLAVE_ADDR,
+						   UC_BACKGROUND_OPCODE,
+						   write_buf, 1,
+						   error)) {
+		g_prefix_error (error, "failed update finish cmd: ");
+		fu_rts54hub_rtd21xx_device_exit_background (self, error);
+		return FALSE;
+	}
+
+	/* exit background-fw mode */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_set_progress (device, 0);
+	if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error)) {
+		fu_rts54hub_rtd21xx_device_exit_background (self, error);
+		return FALSE;
+	}
+
+	write_buf[0] = ISP_BG_CMD_FW_UPDATE_EXIT;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_ISP_SLAVE_ADDR,
+						   UC_BACKGROUND_OPCODE,
+						   write_buf, 1,
+						   error)) {
+		g_prefix_error (error, "exit background-fw mode: ");
+		return FALSE;
+	}
+
+	/* the device needs some time to restart with the new firmware before
+	* it can be queried again */
+	fu_device_sleep_with_progress (device, 1);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_write_firmware_foreground (FuDevice *device,
+						      FuFirmware *firmware,
+						      FwupdInstallFlags flags,
+						      GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	const guint8 *fwbuf;
+	gsize fwbufsz = 0;
+	guint32 project_addr;
+	guint8 project_id_count;
+	guint8 read_buf[10] = { 0x0 };
+	guint8 write_buf[ISP_PACKET_SIZE] = { 0x0 };
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new (self, error);
+	if (locker == NULL) {
+		fu_rts54hub_rtd21xx_device_exit_foreground (self, error);
 		return FALSE;
+	}
 
 	/* simple image */
 	fw = fu_firmware_get_bytes (firmware, error);
-	if (fw == NULL)
+	if (fw == NULL) {
+		fu_rts54hub_rtd21xx_device_exit_foreground (self, error);
```

Comment:
can't this be done on device close instead?

Reply:
I'm very unclear what all the foreground/background stuff does TBH.

Reply:
hi @hughsie ， Background/foreground using different Command int the process of fw update.